### PR TITLE
chore: use new mongodb-runner in place of custom test server script COMPASS-6942

### PR DESCRIPTION
**Pull request by addaleax**
### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [x] Misc

## Types of changes

- [ ] *Backport Needed*
- [x] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-testserver/package.json [hunk 1]
```diff
@@ -0,0 +1,74 @@
+{
+  "name": "@mongodb-js/compass-testserver",
+  "productName": "compass-testserver Plugin",
```
Comment (Anemy) [inline packages/compass-testserver/package.json:3]:
```suggestion
  "productName": "compass-testserver",
```

**Review by Anemy — COMMENTED** [review]
Small thing, should the new package be named `compass-test-server` to keep the convention? Or is there a reason for `compass-testserver`?

*Ah sorry for commenting on a draft!

File: packages/compass-crud/src/stores/crud-store.spec.ts [hunk 2]
```diff
@@ -117,8 +111,9 @@ function waitForState(store, cb, timeout) {
 }
 
 describe('store', function () {
-  this.timeout(5000);
+  this.timeout(30_000);
```
Comment (gribnoysup) [inline packages/compass-crud/src/stores/crud-store.spec.ts:114]: This was here since forever (weirdly enough it's smaller than our default timeout), so not really part of the PR that needs to be addressed, but maybe would make sense to just define this once in .mocharc file? Seems like we set it to 30_000 in a few test files

File: packages/compass-e2e-tests/tests/oidc.test.ts [hunk 3]
```diff
@@ -83,7 +73,6 @@ describe('OIDC integration', function () {
       }
     }
 
-    // TODO(MONGOSH-1306): Get rid of all the setup code to download mongod here... :(
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/tests/oidc.test.ts:86]: We probably still don't want to set up a separate server that doesn't match the one we are supposedly running tests against inside the test setup here. Probably worth creating a different ticket for that?

File: packages/compass-testserver/package.json (hunk 1 above)
Comment (gribnoysup) [reply packages/compass-testserver/package.json:3]: Looks like create-workspace template is adding this by mistake, productName is only "needed" for plugins

File: packages/data-service/src/connect-mongo-client.spec.ts [hunk 4]
```diff
@@ -38,21 +48,11 @@ describe('connectMongoClient', function () {
   });
 
   describe('local', function () {
-    before(function () {
-      if (process.env.EVERGREEN_BUILD_VARIANT === 'rhel') {
-        // TODO: COMPASS-4866
-        console.warn(
-          'test suites using mongodb-runner are flaky on RHEL, skipping'
-        );
-        return this.skip();
```
Comment (gribnoysup) [inline packages/data-service/src/connect-mongo-client.spec.ts:47]: 🎉

File: packages/compass-testserver/package.json (hunk 1 above)
Comment (addaleax) [reply packages/compass-testserver/package.json]: Yeah, thanks for the pointer, removed this line :+1:

File: packages/compass-crud/src/stores/crud-store.spec.ts (hunk 2 above)
Comment (addaleax) [reply packages/compass-crud/src/stores/crud-store.spec.ts]: Fwiw, I bumped it up to 2 minutes here because the server download in CI takes a while (which, yes, probably fair to mention as a disadvantage of this model).

I’m okay with adding it to a shared rc file, yes :+1:

File: packages/compass-e2e-tests/tests/oidc.test.ts (hunk 3 above)
Comment (addaleax) [reply packages/compass-e2e-tests/tests/oidc.test.ts:86]: I get that it’s not ideal, but I’m honestly not quite sure what a better way is. I’ve thought about using the same model that we use in mongosh (and other Compass packages in this PR), i.e. spawning servers at the sites where they are being used, and discussed it with @lerouxb but that’s quite a departure from how compass-e2e-tests currently works (expects setup to be separate from mocha hooks, expects ability to run against existing server, etc.)

So I’m going a bit back and forth here myself…

File: packages/compass-crud/src/stores/crud-store.spec.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-crud/src/stores/crud-store.spec.ts]: Yeah, I get that, I think it's okay, this happens anyway, just in the pretest script currently, I think it's fine if we bump the timeouts for this inside the tests.

I was actually thinking more like adding this to the .mongoshrc file in compass-crud, but yeah, I guess if this will apply to a lot of test suites as probably a lot of them use server it might make sense to even stick it in the shared one. Alternatively maybe this is something that the mocha helper that you added can do? This is kinda going on a tangent, but looking at how this is used in the tests and from some common mocha patterns, maybe it can be changed a little to work sort of like this:

```ts
// mocha helper
export funciton setupTestServer(opts) {
  // setting high timeout because we are anticipating that starting cluster from
  // tests will take more time than our default timeout allows for
  this.timeout(60_000);
  beforeEach(async function() {
    this.cluster = await startTestServer(opts);
  });
  afterEach(async function () {
    if (opts.alwaysStartNewServer) await cluster.stop();
  })
  after(async function () {
    if (!opts.alwaysStartNewServer) await cluster.stop();
  })
}

// some-sort-of-test.spec.ts
describe('when connected to mongodb', function() {
  setupTestServer();

  let client

  before(async function () {
    client = new MongoClient(this.cluster.connectionString);
    await client.connect();
  })
})
```

That way all the test server logic that we have to repeat every time stays in one place, so chances of forgetting to add a timeout or forget the clean up are nonexistent and we can also document things like increased timeout times in one place

Comment (gribnoysup) [reply packages/compass-crud/src/stores/crud-store.spec.ts]: (no strong preference though, just a suggestion!)

Comment (addaleax) [reply packages/compass-crud/src/stores/crud-store.spec.ts]: Yeah no, this makes sense, and it’s ultimately what we ended up doing in mongosh as well :+1: 04b46470dd8eebe92b48d968cc19bfcd49a8eba4 contains this.

File: packages/compass-e2e-tests/tests/oidc.test.ts (hunk 3 above)
Comment (addaleax) [reply packages/compass-e2e-tests/tests/oidc.test.ts:86]: I guess we could, but it would make the code more complex and less local … so I think it would be good to know what the benefits are (starting and stopping servers is ultimately actually fairly fast, it’s only download and extraction that may take a while)

File: packages/compass-test-server/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    await cluster.close();
+    throw new Error('Cluster started while mocha tests were finishing');
+  }
+  clusters.set(key, cluster);
+  return cluster;
+}
+
+// mocha-ized variant of startTestServer(), convenient
+// for automatic teardown at the right scope and integrated timeout
+export function mochaTestServer(
```
Comment (lerouxb) [inline packages/compass-test-server/src/index.ts:38]: nice

File: packages/data-service/src/connect-mongo-client.spec.ts [hunk 6]
```diff
@@ -1,13 +1,16 @@
 import assert from 'assert';
 import asyncHooks from 'async_hooks';
 import { expect } from 'chai';
-import type { CommandStartedEvent } from 'mongodb';
+import type { MongoClientOptions } from 'mongodb';
+import { type CommandStartedEvent } from 'mongodb';
```
Comment (lerouxb) [inline packages/data-service/src/connect-mongo-client.spec.ts:5]: Probably a nit: What's with this style? Can't CommandStartedEvent be imported along with MongoClientOptions in the previous line?

Comment (addaleax) [reply packages/data-service/src/connect-mongo-client.spec.ts:5]: Huh, I’m surprised myself :slightly_smiling_face: This was just what VSCode automatically did for me when I added the usage of `MongoClientOptions`, not something I intentionally added. Yes, they should be merged imo.
